nexus/src/sagas.rs

```diff
@@ -184,12 +187,56 @@ pub fn saga_instance_create() -> SagaTemplate<SagaInstanceCreate> {
             sic_create_network_interfaces_undo,
         ),
     );
+
     template_builder.append(
         "network_interfaces",
         "CreateNetworkInterfaces",
         new_action_noop_undo(sic_create_network_interfaces),
     );
 
+    // Saga actions must be atomic - they have to fully complete or fully abort.
+    // This is because Steno assumes that the saga actions are atomic and
+    // therefore undo actions are *not* run for the failing node.
+    //
+    // For this reason, each disk is created and attached with a separate saga
+    // node. If a saga node had a loop to attach or detach all disks, and one
+    // failed, any disks that were attached would not be detached because the
+    // corresponding undo action would not be run. Separate each disk create and
+    // attach to their own saga node and ensure that each function behaves
+    // atomically.
+    //
+    // Currently, instances can have a maximum of 8 disks attached. Create two
+    // saga nodes for each disk that will unconditionally run but contain
+    // conditional logic depending on if that disk index is going to be used.
+    // Steno does not currently support the saga node graph changing shape.
+    for i in 0..8 {
+        template_builder.append(
```

Comment:
WDYT about calling `append_parallel` here, and running these concurrently? Same for attach?

Reply:
I can't, unfortunately, unless I store the strings somewhere persistent. append_parallel wants tuples that look like `(&str, &str, Arc<dyn steno::Action<SagaInstanceCreate>>)`.

Reply:
Would that not be possible through:

```rust
template_builder.append_parallel(
  (0..MAX_DISKS).map(|i| {
    (
      &format!("create_disks{}", i),
      "CreateDisksForInstance",
      AsyncFunc::new_action(....),
    )
  }).collect::<Vec<_>>()
);
```

(same for "AttachDisks" below)

If Rust gives you issues with lifetimes, then maybe:

```rust
let names = (0...MAX_DISKS).map(|i| format!("create_disks{}", i)).collect::<Vec<String>>();
template_builder.append_parallel(
  (0..MAX_DISKS).map(|i| {
    (
      names[i],
      "CreateDisksForInstance",
      AsyncFunc::new_action(....),
    )
  }).collect::<Vec<_>>()
);
```

Reply:
Rust does give you issues with lifetimes for the first one, and for the second I'm pretty sure `names` has to live longer than the template builder function.